packages/backend/src/services/UserService.ts

```diff
@@ -121,6 +135,45 @@ export class UserService {
         });
     }
 
+    async loginWithOpenId(
+        openIdUser: OpenIdUser,
+        sessionUser: SessionUser | undefined,
+    ): Promise<SessionUser | OpenIdUser> {
+        // Login to existing linked identity
+        try {
+            const loginUser = await this.userModel.getSessionUserByOpenId(
+                openIdUser.openId.issuer,
+                openIdUser.openId.subject,
+            );
+            await this.openIdIdentityModel.updateIdentityByOpenId(
+                openIdUser.openId,
+            );
+            return loginUser;
+        } catch (e) {
+            if (!(e instanceof NotFoundError)) {
+                throw e;
+            }
+        }
+        // Link openid identity to logged-in user
+        if (sessionUser?.userId) {
+            await this.openIdIdentityModel.createIdentity({
+                userId: sessionUser.userId,
+                issuer: openIdUser.openId.issuer,
+                subject: openIdUser.openId.subject,
+                email: openIdUser.openId.email,
+            });
+            return sessionUser;
+        }
+        // Return an OpenId User
+        return openIdUser;
+    }
```

Comment:
Took me some time to understand that this logic depended on not throwing error. 
Imo it would be easier to read if we had a `findSessionUserByOpenId` that doesn't throw if there isn't a match. If a user is returned we would update the identity if not we would link identity to logged in user.

Reply:
If there isn't a user with that open id and userSession is undefined shouldn't we throw an error instead of returning the original openIdUser ?

Reply:
> findSessionUserByOpenId that doesn't throw if there isn't a match

This would be much cleaner thanks!

> shouldn't we throw an error instead of returning the original openIdUser

Yes. Eventually it'll be valid to return `openIdUser` because `req.user` will have type `req.user: SessionUser | OpenIdUser` in order to allow users to authenticate with an external provider while they complete their Lightdash account setup.

Reply:
Refactored this and it also no longer returns an `openIdUser` until we need that functionality